### PR TITLE
Warn when UID is out of local ID ranges

### PR DESCRIPTION
**Pull request by duzda**
Provides simple warning when creating new user with uid out of all local ranges, as this is the main culprit of breaking Kerberos, by not generating ipantsecurityidentifier. We don't have to check for user-mod, because modification never changes ipantsecurityidentifier. We do not have to check groups, as groups are ignored for ipa without AD trust. It's reasonable to revisit this in the future for group creation and warn against groups out of ranges as well as warn for users with groups without SID, in case AD trust is enabled.

Fixes: https://pagure.io/freeipa/issue/9781

## Summary by Sourcery

Add a warning mechanism for user creation when the UID number is outside of local ID ranges, which can prevent proper Kerberos security identifier generation

New Features:
- Implement a check to warn when a user's UID is outside local ID ranges

Bug Fixes:
- Prevent potential Kerberos security identifier generation issues by providing early warning

Tests:
- Add test case to verify warning is thrown when UID is out of range

### CODE REVIEW

**Comment by duzda** [comment]
Also, there has been a proposal of throwing an error instead and provide a flag, that allows to ignore the error. This makes sense, as it's the same behavior as when it comes to deleting idranges, however I believe, if someone is using the `--uid` option, they are well aware of what is going on.

**Comment by rcritten** [comment]
> Also, there has been a proposal of throwing an error instead and provide a flag, that allows to ignore the error. This makes sense, as it's the same behavior as when it comes to deleting idranges, however I believe, if someone is using the `--uid` option, they are well aware of what is going on.

But if this is

File: ipalib/messages.py [hunk 1]
```diff
@@ -519,6 +519,18 @@ class ServerUpgradeRequired(PublicMessage):
...
 
+class UidNumberOutOfLocalIDRange(PublicMessage):
+    """
+    **13034** UID Number is out of all local ID Ranges
+    """
+    errno = 13034
+    type = "warning"
+    format = _(
+        "User '%(user)s', with UID Number '%(uidnumber)d' is out of all ID "
+        "Ranges, 'ipantsecurityidentifier' will not be correctly generated."
```
Comment (rcritten) [inline ipalib/messages.py:530]: I would replace ipantsecurityidentifier with SID. Going into the implications of no SID is probably beyond this but I guess be prepared for the doc folks to ask.

File: ipaserver/plugins/user.py [hunk 2]
```diff
@@ -174,6 +174,27 @@ def check_last_member(user, protected_group_name=u'admins'):
         raise errors.LastMemberError(key=user, label=_(u'group'),
             container=protected_group_name)
 
+
+def is_in_local_idrange(command, uidnumber, options):
+    # Guaranteed, it will be autogenerated by sidgen
```
Comment (rcritten) [inline ipaserver/plugins/user.py:179]: I find the comment misleading. In this case "it" is a SID but the function is only checking that the uidnumber is within an IPA range which only indirectly applies to SID in this case. Someone in the future might use this for something else and the comment could be confusing.

File: ipaserver/plugins/user.py [hunk 3]
```diff
@@ -174,6 +174,27 @@ def check_last_member(user, protected_group_name=u'admins'):
...
+    if uidnumber == -1:
+        return True
+
+    result = command.api.Command.idrange_find(
+        version=options['version'],
+        iparangetype='ipa-local',
+    )
```
Comment (rcritten) [inline ipaserver/plugins/user.py:186]: I wonder if someone would create so many ranges that some results would be truncated. Probably not but it wouldn't hurt to check.

File: ipaserver/plugins/user.py (hunk 2 above)
Comment (duzda) [reply ipaserver/plugins/user.py:179]: You're right, I will handle this outside of the function.

**Review by flo-renaud — COMMENTED** [review]
Hi @duzda 
thanks for the PR, please find my inline comments.

I was also wondering about group-add command. It supports `--gid=GIDNUMBER`, shouldn't we also check for the ranges for this command?

File: ipaserver/plugins/user.py [hunk 4]
```diff
@@ -174,6 +174,24 @@ def check_last_member(user, protected_group_name=u'admins'):
...
+        version=options['version'],
+        iparangetype='ipa-local',
+        sizelimit=0,
+    )
+
+    for r in result['result']:
+        if 'ipabaserid' in r:
+            ipabaseid = int(r['ipabaseid'][0])
+            ipaidrangesize = int(r['ipaidrangesize'][0])
+            if ipabaseid <= uidnumber <= ipabaseid + ipaidrangesize:
```
[the remark this answers is not in the document]
Comment (duzda) [reply ipaserver/plugins/user.py:189]: Of course it should, I wonder why I overlooked this

File: ipaserver/plugins/user.py [hunk 5]
```diff
@@ -174,6 +174,24 @@ def check_last_member(user, protected_group_name=u'admins'):
         raise errors.LastMemberError(key=user, label=_(u'group'),
             container=protected_group_name)
 
+
+def is_in_local_idrange(command, uidnumber, options):
+    result = command.api.Command.idrange_find(
+        version=options['version'],
```
[the remark this answers is not in the document]
Comment (duzda) [reply ipaserver/plugins/user.py:180]: Done, confused me, that it's being used often

File: ipaserver/plugins/user.py [hunk 6]
```diff
@@ -670,6 +688,17 @@ def pre_callback(self, ldap, dn, entry_attrs, attrs_list, *keys, **options):
...
+        # Skip dynamically assigned uid, old clients say 999
+        uidnumber = entry_attrs.get('uidnumber')
+        if (
+            uidnumber != -1
+            and uidnumber != 999
+            and not is_in_local_idrange(self, uidnumber, options)
+        ):
+            self.add_message(messages.UidNumberOutOfLocalIDRange(
+                user=entry_attrs.get('uid'), uidnumber=uidnumber))
+
```
[the remark this answers is not in the document]
Comment (duzda) [reply ipaserver/plugins/user.py:701]: Thanks, moved to the baseuser class

**Comment by flo-renaud** [comment]
There are new errors in the test_xmlrpc/test_stageuser_plugin.py:
```
[2025-05-22 12:20:59] [31mERROR[0m test_xmlrpc/test_stageuser_plugin.py::[1mTestStagedUser::test_create_with_attr[group ID number][0m - AssertionError: did not raise!
[2025-05-22 12:20:59] [31mERROR[0m test_xmlrpc/test_stageuser_plugin.py::[1mTestStagedUser::test_create_with_attr[password][0m - AssertionError: did not raise!
[
```
and
```
[2025-05-22 12:20:59] [31mFAILED[0m test_xmlrpc/test_stageuser_plugin.py::[1mTestStagedUser::test_create_with_attr[user ID number][0m - AssertionError: assert_deepequal: dict keys mismatch.
[2025-05-22 12:20:59] [31mFAILED[0m test_xmlrpc/test_stageuser_plugin.py::[1mTestStagedUser::test_create_with_attr[UID and GID numbers][0m - AssertionError: assert_deepequal: dict keys mismatch.
[
```

**Comment by duzda** [comment]
Should be fixed, I'm definitely open to some suggestions regarding the staged users tests.

**Review by flo-renaud — COMMENTED** [review]
Hi @duzda 

The patch looks good to me, I have a suggestion for an additional test.
Also can you open a separate pagure ticket for checking that gid on group-add is inside the local ranges?
Thanks

**Comment by duzda** [comment]
Thanks, I have created two new issues, https://pagure.io/freeipa/issue/9795 and https://pagure.io/freeipa/issue/9796. The goal is similar for both of them, however they touch different parts of code and imo should be addressed separately.

**Comment by flo-renaud** [comment]
Thanks for the update, the new test LGTM.

**Comment by flo-renaud** [comment]
master:

* b36c163fe8c225e12737d0e25092bb1a7fc9fd5c Warn when UID is out of local ID ranges
